two_factor/views/core.py

```diff
@@ -103,7 +107,12 @@ def done(self, form_list, **kwargs):
         """
         Login the user and redirect to the desired page.
         """
-        login(self.request, self.get_user())
+        # if no token is required, populate user.otp_device
+        # so OTPRequiredMixin will grant access
+        user = self.get_user()
+        if not self.token_required(self.request):
+            user.otp_device = self.get_device()
```

Comment:
I'd rather not specify the user's default device here, but a fake placeholder device instead. This makes it possible to distinguish whether the user actually verified a second factor in this session.